File: question/leetcode/leetcode0416.Partition-Equal-Subset-Sum/leetcode0416.go
package leetcode

func canPartition(nums []int) bool {
	total := 0
	for _, val := range nums {
		total += val
	}
	if total%2 != 0 {
		return false
	}
	target := total >> 1
	dp := make([]bool, target+1)
	dp[0] = true

	for _, num := range nums {
		for i := target; i >= num; i-- {
			dp[i] = dp[i] || dp[i-num]
		}
		if dp[target] {
			return true
		}
	}
	return dp[target]
}

File: question/leetcode/leetcode0416.Partition-Equal-Subset-Sum/leetcode0416.ts
function canPartition(nums: number[]): boolean {
	let vals: number = 0;
	for (const num of nums) {
		vals += num;
	}

	if (vals % 2 !== 0) {
		return false;
	}

	const target = vals >> 1;
	const dp: boolean[] = Array(target + 1).fill(false);
	dp[0] = true;

	for (const num of nums) {
		for (let i = target; i >= num; i--) {
			//如果原来能组成i，或者现在能通过 i -  num + num 组成i，都算成功
			dp[i] = dp[i] || dp[i - num];
			//写成dp[i] = dp[i - num],会覆盖掉之前其他num得到的true
		}

		if (dp[target]) {
			return true
		}
	}

	return dp[target];
};